llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -7611,6 +7613,117 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
   return nullptr;
 }
 
+enum class SignType {
+  Positive,
+  NonPositive,
+  Negative,
+  NonNegative,
+};
+
+/// Check signess of a constant integer or vector of integers
+///
+/// \param C constant to check for signedness
+/// \param SignType the sign type to check against
+///
+/// \return whether constant is signess corresponds with the requesed requested
+/// sign
+static bool checkConstantSignType(const Constant *C, SignType Sign) {
+  auto CheckSign = [Sign](const APInt &Value) {
+    switch (Sign) {
+    case SignType::Positive:
+      return Value.isStrictlyPositive();
+    case SignType::NonPositive:
+      return Value.isNonPositive();
+    case SignType::Negative:
+      return Value.isNegative();
+    case SignType::NonNegative:
+      return Value.isNonNegative();
+    default:
```

Comment:
/home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7641:5: error: default label in switch which covers all enumeration values [-Werror,-Wcovered-switch-default]

Reply:
This helper is generally unnecessary, you can directly use `m_NonNegative()` etc instead.

Reply:
It is used to support non-spalt vector constants. See https://github.com/llvm/llvm-project/pull/155501?notification_referrer_id=NT_kwDOAO7OmbQxODQ5MDk1OTY0OToxNTY1MDQ1Nw#discussion_r2311989751.

Reply:
The m_NonNegative() etc matchers that do not accept an argument also match non-splat vectors.